server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -178,34 +183,88 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
+                                Locale.ROOT,
+                                "required pipeline [%s] and default pipeline [%s] can not both be set",
+                                requiredPipeline,
+                                defaultPipeline);
+                            throw new IllegalArgumentException(message);
+                        }
+                        final String pipeline;
+                        if (requiredPipeline != null) {
+                            pipeline = requiredPipeline;
+                        } else {
+                            pipeline = Objects.requireNonNullElse(defaultPipeline, IngestService.NOOP_PIPELINE_NAME);
+                        }
+                        indexRequest.setPipeline(pipeline);
+                    }
+
+                    if (requestPipeline != null) {
+                        if (requestCanOverridePipeline == false) {
+                            final String message = String.format(
+                                Locale.ROOT,
+                                "request pipeline [%s] can not override required pipeline [%s]",
+                                requestPipeline,
+                                requiredPipeline);
+                            throw new IllegalArgumentException(message);
+                        } else {
+                            indexRequest.setPipeline(requestPipeline);
                         }
                     }
-                } else if (IngestService.NOOP_PIPELINE_NAME.equals(pipeline) == false) {
+
+                    if (IngestService.NOOP_PIPELINE_NAME.equals(indexRequest.getPipeline()) == false) {
+                        hasIndexRequestsWithPipelines = true;
+                    }
+                    if (clusterService.localNode().isIngestNode() == false) {
+                        /*
+                         * We have to track whether or not the pipeline for this request has already been derived. It can happen that the
+                         * pipeline for this request has already been derived yet we execute this loop again. That occurs if the bulk
+                         * request has been forwarded by a non-ingest coordinating node to an ingest node. In this case, the coordinating
+                         * node will have already derived the pipeline for this request. It is important that we are able to distinguish
+                         * this situation as we can not double-derive the pipeline because we will not be able to distinguish the case of
+                         * the pipeline having been set from a request pipeline parameter versus having been set by the derivation. We need
+                         * to be able to distinguish these cases as we need to reject the request if the pipeline was set by a required
+                         * pipeline and there is a request pipeline parameter too.
+                         */
+                        indexRequest.isForwardedRequest(true);
```

Comment:
I wonder if we should flag an index request all time whether the pipeline has been resolved.
In the case the coordinating node is also an ingest node then this logic gets executed twice,
but then the requestPipeline is set to `_none` and this can fail the request if required 
pipeline is used. I think this is the reason why `240_required_pipeline.yml` fails: 
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/6992/

Reply:
Yes, I pushed 01416c17f73ddb970412a274399362fff5122b21.

Reply:
Cool this passes for me now:

```
./gradlew ':modules:ingest-common:integTestRunner' --tests "org.elasticsearch.ingest.common.IngestCommonClientYamlTestSuiteIT.test {yaml=ingest/240_required_pipeline/Test index with required pipeline}" --parallel
```